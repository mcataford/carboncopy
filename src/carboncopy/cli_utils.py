import inquirer

from typing import List
from pathlib import Path


def prompt_staging_files_confirmation(
    staged: List[Path], template_repository_name: str
):
    questions = [
        inquirer.Checkbox(
            "suggested_changes",
            message="The following files can be pulled from {}. Select the ones to be merged in:".format(
                template_repository_name
            ),
            choices=[str(staged_file) for staged_file in staged],
        )
    ]

    return inquirer.prompt(questions)
