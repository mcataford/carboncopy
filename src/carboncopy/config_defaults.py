CONFIG_DEFAULTS = {
    # This is used as the scratch space where the template is cloned.
    # It is cleaned once done.
    "temp_directory": ".carboncopy",
    "ignore": [],
}
