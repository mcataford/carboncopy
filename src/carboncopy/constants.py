RCFILE_PATH = ".carboncopyrc"

FORCED_IGNORE_PATTERNS = r"^.git/*"
FETCH_URL_PATTERN = r"Fetch URL"
GIT_LINK_PATTERN = r"(?<=git@github.com:)[A-Za-z0-9_\-\.]+/[A-Za-z0-9_\-\.]+"
GIT_EXT_PATTERN = r"\.git$"
