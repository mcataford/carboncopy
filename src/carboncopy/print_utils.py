def pretty_print(message: str):
    print(message)
