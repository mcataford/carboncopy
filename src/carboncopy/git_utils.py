from pathlib import Path
import subprocess
import re

import requests

from .constants import FETCH_URL_PATTERN, GIT_EXT_PATTERN, GIT_LINK_PATTERN


class NotInAGitRepositoryError(Exception):
    pass


class NoTemplateError(Exception):
    pass


def clone_template_head(url: str, destination: Path) -> None:
    _run(
        "git clone {url} {location}".format(url=url, location=destination.resolve()),
        stdout=subprocess.DEVNULL,
    )


def get_local_repository_meta():
    stdout = _run("git remote show origin")
    stdout_split = stdout.decode().split("\n")

    for line in stdout_split:
        if re.search(FETCH_URL_PATTERN, line):
            match = re.search(GIT_LINK_PATTERN, line)
            org, repo = match.group(0).split("/")

            return org, re.sub(GIT_EXT_PATTERN, "", repo)

    return None, None


def get_repo_metadata(owner, repo):
    headers = {"Accept": "application/vnd.github.baptiste-preview+json"}
    r = requests.get(
        "https://api.github.com/repos/{owner}/{repo}".format(owner=owner, repo=repo),
        headers=headers,
    )
    repo_data = r.json()

    template_repo = repo_data.get("template_repository")
    return template_repo


def _run(command, stdout=subprocess.PIPE):
    return subprocess.run(command.split(" "), stdout=stdout, stderr=stdout).stdout
