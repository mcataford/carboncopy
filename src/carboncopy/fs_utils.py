import shutil
from pathlib import Path
from typing import Union, List, Dict, Any
import os

from .print_utils import pretty_print


class Transform:
    def __init__(self, source: Path, destination: Path):
        self.source = source
        self.destination = destination

    def get_source(self, as_str: bool = False):
        return str(self.source) if as_str else self.source

    def get_destination(self, as_str: bool = False):
        return str(self.destination) if as_str else self.destination

    def __repr__(self):
        return "<{classname} {source} -> {destination}>".format(
            classname=self.__class__,
            source=str(self.source),
            destination=str(self.destination),
        )


def clean_temp_files(path: Path) -> None:
    if path:
        shutil.rmtree(path, True)


def get_template_transforms(path: Path) -> List[Transform]:
    file_paths = []

    stack = [path]

    while stack:
        current_path = stack.pop()

        if not current_path.is_dir():
            file_paths.append(current_path)
            continue

        for child in current_path.iterdir():
            child_path = Path(child)
            stack.append(child_path)

    return [
        Transform(source=filename, destination=filename.relative_to(path))
        for filename in file_paths
    ]


def squash(transform: Transform) -> None:
    destination = transform.get_destination()
    source = transform.get_source()

    if not destination.parent.exists():
        os.makedirs(destination.parent)

    try:
        shutil.copy(source, destination)
        pretty_print(
            "Copied {source} -> {destination}".format(
                source=source, destination=destination
            )
        )
    except IsADirectoryError:
        pretty_print(
            "Failed to copy {source} -> {destination}".format(
                source=source, destination=destination
            )
        )
    except Exception as e:
        pretty_print(e.__class__)
        pretty_print(e)
