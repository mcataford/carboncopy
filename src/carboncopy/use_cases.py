import subprocess
import re
from pathlib import Path
import requests
import shutil
import os
import json
import uuid
from typing import List, Dict, Any

from .config_defaults import CONFIG_DEFAULTS
from .constants import RCFILE_PATH, FORCED_IGNORE_PATTERNS
from .fs_utils import Transform, squash, clean_temp_files, get_template_transforms
from .git_utils import (
    NoTemplateError,
    NotInAGitRepositoryError,
    get_local_repository_meta,
    get_repo_metadata,
    clone_template_head,
)
from .cli_utils import prompt_staging_files_confirmation
from .print_utils import pretty_print


def get_local_config(root_path: Path = Path(".")) -> Dict[str, Any]:
    config_path = root_path.joinpath(RCFILE_PATH)
    try:
        with open(config_path, "r") as config_file:
            loaded_config = json.load(config_file)

        merged_config = CONFIG_DEFAULTS.copy()

        for key in loaded_config:
            merged_config[key] = loaded_config.get(key)

        return merged_config

    except FileNotFoundError:
        pretty_print(
            "No config file found in current directory! Proceeding with defaults."
        )
    except ValueError:
        pretty_print("Invalid RC file!")

    return CONFIG_DEFAULTS


class UseCases:
    def __init__(self, config: Dict[str, Any]) -> None:
        self.config = config
        self.template_repo: Dict[str, str] = {}
        self.org = None
        self.repo = None

    def fetch_template_repository_details(self) -> None:
        org, repo = get_local_repository_meta()
        template_repo_data = get_repo_metadata(org, repo)

        if not (org and repo):
            raise NotInAGitRepositoryError()

        if not template_repo_data:
            raise NoTemplateError()

        self.org = org
        self.repo = repo
        self.template_repo = template_repo_data

    def clone_template_repository(self) -> None:
        clone_template_head(
            self.template_repo["clone_url"], Path(self.config["temp_directory"])
        )

    def stage_changes(self) -> List[Transform]:
        path = Path(self.config["temp_directory"])
        available_transforms = get_template_transforms(path)

        def can_stage(path_str: str) -> bool:
            return not re.match(FORCED_IGNORE_PATTERNS, path_str) and all(
                [re.match(patt, path_str) for patt in self.config["ignore"]]
            )

        allowed_transforms = [
            transform
            for transform in available_transforms
            if can_stage(transform.get_destination(as_str=True))
        ]

        destinations = [
            transform.get_destination(as_str=True) for transform in allowed_transforms
        ]
        chosen_files = prompt_staging_files_confirmation(
            destinations, "{}/{}".format(self.org, self.repo)
        )
        chosen_transforms = [
            transform
            for transform in allowed_transforms
            if transform.get_destination(as_str=True)
            in chosen_files["suggested_changes"]
        ]

        return chosen_transforms

    def apply_changes(self, paths: List[Transform]) -> None:
        for path in paths:
            squash(path)

    def clean_up(self) -> None:
        clean_temp_files(Path(self.config["temp_directory"]))
