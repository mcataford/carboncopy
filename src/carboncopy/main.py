from .use_cases import get_local_config, UseCases
from .print_utils import pretty_print
from .git_utils import NoTemplateError, NotInAGitRepositoryError


def run():
    config = get_local_config()
    use_cases = UseCases(config)

    try:
        use_cases.fetch_template_repository_details()
    except NotInAGitRepositoryError:
        pretty_print("Not in a git repository.")
        return 1
    except NoTemplateError:
        pretty_print("This repository does not have a template associated with it.")
        return 1

    try:
        use_cases.clone_template_repository()
        paths = use_cases.stage_changes()
        use_cases.apply_changes(paths)
    except Exception as e:
        pretty_print(e.__class__)
        pretty_print(e)
    finally:
        use_cases.clean_up()
