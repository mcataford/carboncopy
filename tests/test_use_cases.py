import json

import pytest

from src.carboncopy.use_cases import get_local_config, UseCases
from src.carboncopy.constants import RCFILE_PATH
from src.carboncopy.config_defaults import CONFIG_DEFAULTS
from src.carboncopy.git_utils import NotInAGitRepositoryError, NoTemplateError

from .test_utils import assert_captured_output_matches_snapshot

def test_get_local_config_returns_default_config_if_no_config_file_present(tmp_path, snapshot, capsys):
    assert len(list(tmp_path.iterdir())) == 0

    fetched_config = get_local_config(tmp_path)

    assert fetched_config == CONFIG_DEFAULTS

    assert_captured_output_matches_snapshot(capsys, snapshot)

def test_get_local_config_returns_default_config_if_invalid_config_file_present(tmp_path, snapshot, capsys):
    invalid_config_file = tmp_path / RCFILE_PATH

    invalid_config_file.write_text('')

    assert len(list(tmp_path.iterdir())) == 1

    fetched_config = get_local_config(tmp_path)

    assert fetched_config == CONFIG_DEFAULTS

    assert_captured_output_matches_snapshot(capsys, snapshot)

def test_get_local_config_merges_rcfile_with_default_config(tmp_path, snapshot, capsys):
    valid_config = { "ignore": ["some-file.md"] }

    config_file = tmp_path / RCFILE_PATH
    config_file.write_text(json.dumps(valid_config))

    assert len(list(tmp_path.iterdir())) == 1

    fetched_config = get_local_config(tmp_path)

    expected_config = {**CONFIG_DEFAULTS, **valid_config}

    assert fetched_config == expected_config

    assert_captured_output_matches_snapshot(capsys, snapshot) 
