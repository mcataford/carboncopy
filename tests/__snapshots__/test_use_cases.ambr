# name: test_fetch_template_repository_details_throws_NoTemplateError_if_no_template_repo
  ''
---
# name: test_fetch_template_repository_details_throws_NoTemplateError_if_no_template_repo.1
  ''
---
# name: test_fetch_template_repository_details_throws_NotInGitRepositoryError_if_not_in_repo
  ''
---
# name: test_fetch_template_repository_details_throws_NotInGitRepositoryError_if_not_in_repo.1
  ''
---
# name: test_get_local_config_merges_rcfile_with_default_config
  ''
---
# name: test_get_local_config_merges_rcfile_with_default_config.1
  ''
---
# name: test_get_local_config_returns_default_config_if_invalid_config_file_present
  '
  Invalid RC file!
  
  '
---
# name: test_get_local_config_returns_default_config_if_invalid_config_file_present.1
  ''
---
# name: test_get_local_config_returns_default_config_if_no_config_file_present
  '
  No config file found in current directory! Proceeding with defaults.
  
  '
---
# name: test_get_local_config_returns_default_config_if_no_config_file_present.1
  ''
---
# name: test_stage_changes_creates_transforms_for_all_valid_changes
  ''
---
# name: test_stage_changes_creates_transforms_for_all_valid_changes.1
  ''
---
# name: test_stage_changes_ignores_all_configured_ignore_patterns
  ''
---
# name: test_stage_changes_ignores_all_configured_ignore_patterns.1
  ''
---
# name: test_stage_changes_ignores_all_forced_ignore_patterns
  ''
---
# name: test_stage_changes_ignores_all_forced_ignore_patterns.1
  ''
---
